### PR TITLE
Add vector div and scalar ops

### LINKED ISSUE #2942: Implement divide functionality for Vec2, Vec3 and Vec4 classes
Currently we have:
- mul - multiplies vector by vector
- scale - multiplies vector by scalar

We should consider similar functionality for divide operation.

### PR DESCRIPTION
**Pull request by willeastcott**
Adds the following API:

```typescript
Vec2.addScalar(scalar: number) : Vec2
Vec2.div(v: Vec2) : Vec2
Vec2.div2(v1: Vec2, v2: Vec2) : Vec2
Vec2.divScalar(scalar: number) : Vec2
Vec2.mulScalar(scalar: number) : Vec2
Vec2.subScalar(scalar: number) : Vec2

Vec3.addScalar(scalar: number) : Vec3
Vec3.div(v: Vec3) : Vec3
Vec3.div2(v1: Vec3, v2: Vec3) : Vec3
Vec3.divScalar(scalar: number) : Vec3
Vec3.mulScalar(scalar: number) : Vec3
Vec3.subScalar(scalar: number) : Vec3

Vec4.addScalar(scalar: number) : Vec4
Vec4.div(v: Vec4) : Vec4
Vec4.div2(v1: Vec4, v2: Vec4) : Vec4
Vec4.divScalar(scalar: number) : Vec4
Vec4.mulScalar(scalar: number) : Vec4
Vec4.subScalar(scalar: number) : Vec4
```

Deprecates the following API:

```typescript
Vec2.scale(scale: number) : Vec2
Vec3.scale(scale: number) : Vec3
Vec4.scale(scale: number) : Vec4
```

Adds unit tests for all new API.

Fixes #2942 

I confirm I have signed the [Contributor License Agreement](https://docs.google.com/a/playcanvas.com/forms/d/1Ih69zQfJG-QDLIEpHr6CsaAs6fPORNOVnMv5nuo0cjk/viewform).

### CODE REVIEW

File: src/math/vec2.js [hunk 1]
```diff
@@ -164,6 +185,74 @@ class Vec2 {
...
+     * the specified vector.
+     * @param {Vec2} lhs - The dividend vector (the vector being divided).
+     * @param {Vec2} rhs - The divisor vector (the vector dividing the dividend).
+     * @returns {Vec2} Self for chaining.
+     * @example
+     * var a = new pc.Vec2(4, 9);
+     * var b = new pc.Vec2(2, 3);
+     * var r = new pc.Vec2();
+     *
+     * r.add2(a, b);
```
Comment (slimbuck) [inline src/math/vec2.js:223]: div2

File: src/math/vec2.js [hunk 2]
```diff
@@ -297,6 +386,27 @@ class Vec2 {
...
+    /**
+     * @function
+     * @name Vec2#mulScalar
+     * @description Multiplies each element of a vector by a number.
+     * @param {number} scalar - The number to multiply by.
+     * @returns {Vec2} Self for chaining.
+     * @example
+     * var vec = new pc.Vec2(3, 6);
+     *
+     * vec.divScalar(3);
```
Comment (slimbuck) [inline src/math/vec2.js:398]: mulScalar

File: src/math/vec2.js [hunk 3]
```diff
@@ -297,6 +386,27 @@ class Vec2 {
...
     }
 
+    /**
+     * @function
+     * @name Vec2#mulScalar
+     * @description Multiplies each element of a vector by a number.
+     * @param {number} scalar - The number to multiply by.
+     * @returns {Vec2} Self for chaining.
+     * @example
+     * var vec = new pc.Vec2(3, 6);
```
Comment (slimbuck) [inline src/math/vec2.js:396]: should these be using let?

File: src/math/vec2.js [hunk 4]
```diff
@@ -415,6 +499,27 @@ class Vec2 {
...
+     * @description Subtracts a number from each element of a vector.
+     * @param {number} scalar - The number to subtract.
+     * @returns {Vec2} Self for chaining.
+     * @example
+     * var vec = new pc.Vec2(3, 4);
+     *
+     * vec.subScalar(2);
+     *
+     * // Should output [1, 2]
+     * console.log("The result of the subtraction is: " + a.toString());
```
Comment (slimbuck) [inline src/math/vec2.js:514]: vec.toString()

File: src/math/vec3.js [hunk 5]
```diff
@@ -109,6 +109,28 @@ class Vec3 {
...
+     * @description Adds a number to each element of a vector.
+     * @param {number} scalar - The number to add.
+     * @returns {Vec3} Self for chaining.
+     * @example
+     * var vec = new pc.Vec3(3, 4, 5);
+     *
+     * vec.addScalar(2);
+     *
+     * // Should output [5, 6, 7]
+     * console.log("The result of the addition is: " + a.toString());
```
Comment (slimbuck) [inline src/math/vec3.js:124]: vec.toString()

File: src/math/vec3.js [hunk 6]
```diff
@@ -193,6 +215,77 @@ class Vec3 {
...
+     * the specified vector.
+     * @param {Vec3} lhs - The dividend vector (the vector being divided).
+     * @param {Vec3} rhs - The divisor vector (the vector dividing the dividend).
+     * @returns {Vec3} Self for chaining.
+     * @example
+     * var a = new pc.Vec3(4, 9, 16);
+     * var b = new pc.Vec3(2, 3, 4);
+     * var r = new pc.Vec3();
+     *
+     * r.add2(a, b);
```
Comment (slimbuck) [inline src/math/vec3.js:254]: div2

File: src/math/vec3.js [hunk 7]
```diff
@@ -193,6 +215,77 @@ class Vec3 {
...
+     * @description Divides each element of a vector by a number.
+     * @param {number} scalar - The number to divide by.
+     * @returns {Vec3} Self for chaining.
+     * @example
+     * var vec = new pc.Vec3(3, 6, 9);
+     *
+     * vec.divScalar(3);
+     *
+     * // Should output [1, 2, 3]
+     * console.log("The result of the division is: " + a.toString());
```
Comment (slimbuck) [inline src/math/vec3.js:279]: vec.toString()

File: src/math/vec3.js [hunk 8]
```diff
@@ -329,6 +422,28 @@ class Vec3 {
...
+    /**
+     * @function
+     * @name Vec3#mulScalar
+     * @description Multiplies each element of a vector by a number.
+     * @param {number} scalar - The number to multiply by.
+     * @returns {Vec3} Self for chaining.
+     * @example
+     * var vec = new pc.Vec3(3, 6, 9);
+     *
+     * vec.divScalar(3);
```
Comment (slimbuck) [inline src/math/vec3.js:434]: mulScalar

File: src/math/vec3.js [hunk 9]
```diff
@@ -329,6 +422,28 @@ class Vec3 {
...
+     *
+     * // Should output [9, 18, 27]
+     * console.log("The result of the multiplication is: " + a.toString());
```
Comment (slimbuck) [inline src/math/vec3.js:437]: vec.toString()

File: src/math/vec4.js [hunk 10]
```diff
@@ -168,6 +191,80 @@ class Vec4 {
...
+     * the specified vector.
+     * @param {Vec4} lhs - The dividend vector (the vector being divided).
+     * @param {Vec4} rhs - The divisor vector (the vector dividing the dividend).
+     * @returns {Vec4} Self for chaining.
+     * @example
+     * var a = new pc.Vec4(4, 9, 16, 25);
+     * var b = new pc.Vec4(2, 3, 4, 5);
+     * var r = new pc.Vec4();
+     *
+     * r.add2(a, b);
```
Comment (slimbuck) [inline src/math/vec4.js:231]: div2

File: src/math/vec2.js [hunk 11]
```diff
@@ -93,6 +93,27 @@ class Vec2 {
...
+     * @name Vec2#addScalar
+     * @description Adds a number to each element of a vector.
+     * @param {number} scalar - The number to add.
+     * @returns {Vec2} Self for chaining.
+     * @example
+     * var vec = new pc.Vec2(3, 4);
+     *
+     * vec.addScalar(2);
+     *
+     * // Should output [5, 6]
```
Comment (slimbuck) [inline src/math/vec2.js:107]: I'd prefer the comment to strike a confident tone and say something like 'outputs [5, 6]', but I understand the existing examples all use this language.

File: src/math/vec2.js (hunk 1 above)
Comment (willeastcott) [reply src/math/vec2.js:223]: For these kinds of changes, please use GitHub's `suggestion` feature, so I can quickly merge them. 😄

File: src/math/vec2.js (hunk 3 above)
Comment (willeastcott) [reply src/math/vec2.js:396]: Yes, we can start thinking about switching to ES6 for JSDoc example code. But not in this PR. 😄

File: src/math/vec2.js (hunk 11 above)
Comment (willeastcott) [reply src/math/vec2.js]: Done. Thanks!

File: src/math/vec2.js (hunk 2 above)
Comment (willeastcott) [reply src/math/vec2.js]: Fixed.

File: src/math/vec3.js (hunk 6 above)
Comment (willeastcott) [reply src/math/vec3.js]: Fixed.

File: src/math/vec3.js (hunk 7 above)
Comment (willeastcott) [reply src/math/vec3.js]: Fixed.

File: src/math/vec3.js (hunk 8 above)
Comment (willeastcott) [reply src/math/vec3.js]: Fixed.

File: src/math/vec3.js (hunk 9 above)
Comment (willeastcott) [reply src/math/vec3.js]: Fixed.

File: src/math/vec4.js (hunk 10 above)
Comment (willeastcott) [reply src/math/vec4.js]: Fixed.

File: src/math/vec3.js (hunk 5 above)
Comment (willeastcott) [reply src/math/vec3.js]: Fixed.

File: src/math/vec2.js (hunk 4 above)
Comment (willeastcott) [reply src/math/vec2.js]: Fixed.

**Comment by willeastcott** [comment]
Thanks for catching all those little mistakes @slimbuck. I feel much better about this PR now! 😄

File: src/math/vec4.js [hunk 12]
```diff
@@ -307,6 +404,29 @@ class Vec4 {
...
+    /**
+     * @function
+     * @name Vec4#mulScalar
+     * @description Multiplies each element of a vector by a number.
+     * @param {number} scalar - The number to multiply by.
+     * @returns {Vec4} Self for chaining.
+     * @example
+     * var vec = new pc.Vec4(3, 6, 9, 12);
+     *
+     * vec.divScalar(3);
```
Comment (slimbuck) [inline src/math/vec4.js:416]:
```suggestion
     * vec.mulScalar(3);
```

File: tests/math/test_vec2.js [hunk 13]
```diff
@@ -195,6 +241,15 @@ describe("pc.Vec2", function () {
         equal(2, r.y);
     });
 
+    it("subScalar", function() {
+        var v = new pc.Vec2(2, 4, 6);
```
Comment (slimbuck) [inline tests/math/test_vec2.js:245]:
```suggestion
        var v = new pc.Vec2(2, 4);
```

**Review by slimbuck — APPROVED** [review]

**Review by mvaligursky — APPROVED** [review]

**Pull request merged**